lib/topological_inventory/ansible_tower/operations/source.rb

```diff
@@ -10,58 +10,113 @@ class Source
 
-        def connection_check(source_id)
-          endpoints = api_client.list_source_endpoints(source_id)&.data || []
-          endpoint = endpoints.find(&:default)
-          return STATUS_UNAVAILABLE unless endpoint
+        def required_params
+          %w[source_id]
+        end
 
-          endpoint_authentications = api_client.list_endpoint_authentications(endpoint.id.to_s).data || []
-          return STATUS_UNAVAILABLE if endpoint_authentications.empty?
+        def params_missing?
+          is_missing = false
+          required_params.each do |attr|
+            if (is_missing = params[attr].blank?)
+              logger.error("Source#availability_check - Missing #{attr} for the availability_check request")
+              break
+            end
+          end
 
-          auth_id = endpoint_authentications.first.id
-          auth = Core::AuthenticationRetriever.new(auth_id, identity).process
-          return STATUS_UNAVAILABLE unless auth
+          is_missing
+        end
+
+        def connection_status
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:endpoint_not_found]] unless endpoint
+          return [STATUS_UNAVAILABLE, ERROR_MESSAGES[:authentication_not_found]] unless authentication
+
+          connection_check
+        end
 
+        def connection_check
           connection = ::TopologicalInventory::AnsibleTower::Connection.new
-          connection = connection.connect(endpoint.host, auth.username, auth.password)
+          connection = connection.connect(endpoint.host, authentication.username, authentication.password)
           connection.api.version
 
-          STATUS_AVAILABLE
+          [STATUS_AVAILABLE, nil]
         rescue => e
           logger.error("Failed to connect to Source id:#{source_id} - #{e.message}")
-          STATUS_UNAVAILABLE
+          [STATUS_UNAVAILABLE, e.message]
+        end
+
+        def update_source_and_endpoint(status, error_message = nil)
+          logger.info("Source#availability_check - updating source [#{source_id}] status [#{status}] message [#{error_message}]")
+
+          update_source(status)
+          update_endpoint(status, error_message)
+        end
+
+        def update_source(status)
+          source = ::SourcesApiClient::Source.new
+          source.availability_status = status
+
+          api_client.update_source(source_id, source)
+        rescue SourcesApiClient::ApiError => e
+          logger.error("Source#availability_check - Failed to update Source id:#{source_id} - #{e.message}")
+        end
+
+        def update_endpoint(status, error_message)
+          if endpoint.nil?
+            logger.error("Source#availability_check - Failed to update Endpoint for Source id:#{source_id}. Endpoint not found")
+            return
+          end
+
+          endpoint_update = SourcesApiClient::Endpoint.new
+
+          endpoint_update.availability_status = status
+          endpoint_update.availability_status_error = error_message # if error_message
```

Comment:
does this `# if error_message` need to be uncommented? maybe we could just default it to `nil` that way we wouldn't have to check it. 

EDIT:
After further review it looks like it _is_ defaulting to nil! Maybe we can just remove the comment then?